File: SailClasses.ts
class Competition
{
    Name: string;
    StartDate: Date;
    EndDate: Date;
    Races: Array<Race> = new Array<Race>();

    constructor(name: string, start: Date, end: Date)
    {
        this.Name = name;
        this.StartDate = start;
        this.EndDate = end;
    }
}

class Race
{
    RaceID: number;
    RaceDate: Date;
    Results: Array<RaceResult> = new Array<RaceResult>();

    constructor(id: number, date: Date)
    {
        this.RaceID = id;
        this.RaceDate = date;
    }
}

class RaceResult
{
    BoatID: number;
    Rank: number;
    ElapsedTime: number;
    Laps: number;
    CorrectedTime: number;
    Points: number;
    SpecialMentions: string;

    constructor(id: number, rank: number, elapsed: number, laps: number, corrected: number, points: number, mentions: string)
    {
        this.BoatID = id;
        this.Rank = rank;
        this.ElapsedTime = elapsed;
        this.Laps = laps;
        this.CorrectedTime = corrected;
        this.Points = points;
        this.SpecialMentions = mentions;
    }
}

class Boat
{
    BoatID: number;
    HelmName: string;
    CrewName: string;
    PY: number;
    BoatClass: string;

    constructor(id: number, helmName: string, crewName: string, py: number, boatClass: string)
    {
        this.BoatID = id;
        this.BoatClass = boatClass;
        this.HelmName = helmName;
        this.CrewName = crewName;
        this.PY = py;
    }
}

File: SailClasses.dart
class Competition {
  String name;
  DateTime startDate;
  DateTime endDate;
  List<Race> races = new List<Race>();

  Competition(String name, DateTime start, DateTime end) {
    this.name = name;
    this.startDate = start;
    this.endDate = end;
  }

  String toString() {
    return "Competion ${this.name} starts on ${this.startDate} and ends on ${this.endDate}. It has ${this.races.length} races";
  }
}

class Boat {
  int boatID;
  String helmName;
  String crewName;
  int pY;
  String boatClass;

  Boat(int id, String helmName, String crewName, int py, String boatClass) {
    this.boatID = id;
    this.helmName = boatClass;
    this.helmName = helmName;
    this.crewName = crewName;
    this.pY = py;
  }

  String toString() {
    return "Boat with id ${this.boatID}, helmed by ${this.helmName}";
  }
}


class Race {
  int raceID;
  DateTime raceDate;
  List<RaceResult> results = new List<RaceResult>();

  Race(int id, DateTime date) {
    this.raceID = id;
    this.raceDate = date;
  }

  String toString() {
    return "This is a Race with id:${this.raceID}, starting on ${this.raceDate.toString()}";
  }
}

class RaceResult {
  int boatID;
  int rank;
  int elapsedTime;
  int laps;
  int correctedTime;
  int points;
  String specialMentions;

  RaceResult(int id, int rank, int elapsed, int laps, int corrected, int points,
      String mentions) {
    this.boatID = id;
    this.rank = rank;
    this.elapsedTime = elapsed;
    this.laps = laps;
    this.correctedTime = corrected;
    this.points = points;
    this.specialMentions = mentions;
  }

  String toString() {
    return "The boat with id ${this.boatID} finished on rank ${this.rank} with ${this.points} points";
  }
}

void main() {
  print(Race(1, DateTime.now()));
}
